### PR TITLE
x64: assert that temporary and destination registers match during renaming

### PR DESCRIPTION
**Pull request by abrown**
@cfallin, here is the change we just discussed to ensure that renaming operates on the same numbers of temporary and destination registers. I see it failing on the `address64.wast` spec test, which would indicate to me that either the assertion is too strong or the memory64 codegen has some hidden issue.

### CODE REVIEW

**Comment by cfallin** [comment]
Thanks @abrown! cc @fitzgen: context is that I was helping Andrew debug some ISLE lowerings and we hit a footgun where he was using `with_flags` rather than `with_flags_1`, returning a `ValueRegs` of zero registers, hence this new assertion. From that,

- We should have this assertion, and we should fix the assertion failure (I can take a look maybe tomorrow if someone else doesn't get there first)
- We should maybe name `with_flags_*` more descriptively... slightly lame suggestions could be `with_flags_return_producer`, `with_flags_return_consumer`, ..., but maybe there are better names?

**Review by fitzgen — COMMENTED** [review]
Okay, yes this assertion makes sense. Not good if this is failing.

**Comment by fitzgen** [comment]
> We should maybe name `with_flags_*` more descriptively... slightly lame suggestions could be `with_flags_return_producer`, `with_flags_return_consumer`, ..., but maybe there are better names?

Maybe without the "return" so it is

* `with_flags_none`: gives zero regs
* `with_flags_producer`: gives the producer result
* `with_flags_consumer`: gives the consumer result
* `with_flags_both`: gives `(producer, consumer)`

?

**Review by fitzgen — COMMENTED** [review]
Could you add this assertion to `isa/aarch64/lower/isle.rs` as well?

Alex is planning on de-duping this stuff, but in the meantime we should make sure we are consistent.

**Comment by cfallin** [comment]
It looks like an issue with `iadd_ifcout`; the lowering only returns the add result, but it's supposed to be both the add result and the carry flag.

I pushed an update to the assertion to [the same branch in my fork](https://github.com/cfallin/wasmtime/tree/assert-tmp-dsts) (it was asserting just for one dst, not for all, but it's still a mismatch). Will look at the lowering, maybe it's simple to fix, otherwise we can just back out `iadd_ifcout` from ISLE for now to allow the release, I think.

**Comment by cfallin** [comment]
Further interesting twists: `iadd_ifcout` in the pre-ISLE world also wrote only to its first output (the add result) when lowered by itself. We have a separate pattern-match when `iadd_ifcout` is used by a `trapif` to explicltly do the right thing. As I recall, this special-casing is just enough to get the dynamic heap checks to work, and since we eventually want `iflags`-type values to be replaced by bools and pattern-matching, I don't think we want to fully reify the `iflags` in ISLE. Probably the best option is to just return a freshly-allocated second temp for the flags for now; anything that uses `iflags` will not be handled by ISLE so we will not be in danger of actually using it.

**Comment by cfallin** [comment]
I pushed a fix to the `iadd_ifcout` lowering [here](https://github.com/cfallin/wasmtime/commit/aed39a1a795fba9fa3e550d4376d1ecb45419471); please feel free to cherry-pick this and [the assertion fix](https://github.com/cfallin/wasmtime/commit/c9e66ce81169de17c30dc45651865f0123c9b8eb)!

File: cranelift/codegen/src/isa/x64/lower.isle [hunk 1]
```diff
@@ -153,36 +153,56 @@
...
+                         (put_in_reg x)
+                         (RegMemImm.Reg (put_in_reg y)))
+                    unused_iflags)))
 
 ;; Add a register and an immediate.
 
 (rule (lower (has_type (fits_in_64 ty)
                        (iadd_ifcout x (simm32_from_value y))))
-      (value_reg (add ty (put_in_reg x) y)))
+      (let ((unused_iflags Reg (writable_reg_to_reg (temp_writable_reg $I64))))
```
Comment (fitzgen) [inline cranelift/codegen/src/isa/x64/lower.isle:179]: Here and elsewhere in this commit:

```suggestion
(let ((unused_iflags Reg (writable_reg_to_reg (temp_writable_reg ty))))
```

Comment (cfallin) [reply cranelift/codegen/src/isa/x64/lower.isle:179]: I'm actually not sure about that; this value is a placeholder for the `iflags`, so the type is logically unconnected to the `iadd`. (The register isn't actually used so it won't e.g. cause any additional pressure in whatever register class, but it seems odd to use `ty` here regardless.)

Comment (fitzgen) [reply cranelift/codegen/src/isa/x64/lower.isle:179]: Ah yes, that makes sense.

**Review by cfallin — APPROVED** [review]
Happy to give an r+ as requested but @fitzgen should probably ~~review~~ (EDIT: confirm / r+) it too since it has my commits included :-)

**Review by fitzgen — APPROVED** [review]
:+1:

**Pull request merged**